go/libraries/doltcore/table/editor/pk_table_editor.go

```diff
@@ -46,11 +46,13 @@ func init() {
 	}
 }
 
+type PKDuplicateErrFunc func(keyString string, k, v types.Tuple) error
+
 type TableEditor interface {
-	InsertKeyVal(ctx context.Context, key, val types.Tuple, tagToVal map[uint64]types.Value) error
+	InsertKeyVal(ctx context.Context, key, val types.Tuple, tagToVal map[uint64]types.Value, errFunc PKDuplicateErrFunc) error
 
-	InsertRow(ctx context.Context, r row.Row) error
-	UpdateRow(ctx context.Context, old, new row.Row) error
+	InsertRow(ctx context.Context, r row.Row, errFunc PKDuplicateErrFunc) error
```

Comment:
Rather than change all these interfaces, just add a DuplicateKeyErrorGenerator as a field to the relevant classes and make it settable. Kind of onerous to pass this around everywhere